akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/PropsAdapter.scala

```diff
@@ -18,31 +20,47 @@ import akka.dispatch.Mailboxes
-      case unknown                       => throw new RuntimeException(s"Unsupported dispatcher selector: $unknown")
-    }).withDeploy(Deploy.local) // disallow remote deployment for typed actors
-
-    val mailboxProps = props.firstOrElse[MailboxSelector](MailboxSelector.default()) match {
-      case _: DefaultMailboxSelector           => dispatcherProps
-      case BoundedMailboxSelector(capacity, _) =>
-        // specific support in classic Mailboxes
-        dispatcherProps.withMailbox(s"${Mailboxes.BoundedCapacityPrefix}$capacity")
-      case MailboxFromConfigSelector(path, _) =>
-        dispatcherProps.withMailbox(path)
-      case unknown => throw new RuntimeException(s"Unsupported mailbox selector: $unknown")
-    }
-
-    val localDeploy = mailboxProps.withDeploy(Deploy.local) // disallow remote deployment for typed actors
-
-    val tags = props.firstOrElse[ActorTags](ActorTagsImpl.empty).tags
-    if (tags.isEmpty) localDeploy
-    else localDeploy.withActorTags(tags)
+    val deploy =
+      if (props eq Props.empty) DefaultTypedDeploy // optimized case with no props specified
+      else {
+        val deployWithMailbox = props.firstOrElse[MailboxSelector](MailboxSelector.default()) match {
+          case _: DefaultMailboxSelector           => DefaultTypedDeploy
+          case BoundedMailboxSelector(capacity, _) =>
+            // specific support in classic Mailboxes
+            DefaultTypedDeploy.copy(mailbox = s"${Mailboxes.BoundedCapacityPrefix}$capacity")
+          case MailboxFromConfigSelector(path, _) =>
+            DefaultTypedDeploy.copy(mailbox = path)
+          case unknown => throw new RuntimeException(s"Unsupported mailbox selector: $unknown")
+        }
+
+        val deployWithDispatcher = props.firstOrElse[DispatcherSelector](DispatcherDefault.empty) match {
+          case _: DispatcherDefault          => deployWithMailbox
+          case DispatcherFromConfig(name, _) => deployWithMailbox.copy(dispatcher = name)
+          case _: DispatcherSameAsParent     => deployWithMailbox.copy(dispatcher = Deploy.DispatcherSameAsParent)
+          case unknown                       => throw new RuntimeException(s"Unsupported dispatcher selector: $unknown")
+        }
+
+        val tags = props.firstOrElse[ActorTags](ActorTagsImpl.empty).tags
+        val deployWithTags =
+          if (tags.isEmpty) deployWithDispatcher
+          else deployWithDispatcher.withTags(tags)
+
+        if (deployWithTags.scope != LocalScope) // only replace if changed, withDeploy is expensive
+          deployWithTags.copy(scope = Deploy.local.scope) // disallow remote deployment for typed actors
+        else deployWithTags
+      }
+
+    // avoid the apply methods and also avoid copying props, for performance reasons
+    new akka.actor.Props(
+      deploy,
+      TypedCreatorFunctionConsumerClazz,
+      ActorAdapterClazz :: (() => new ActorAdapter(behavior(), rethrowTypedFailure)) :: Nil)
```

Comment:
this will create the `_producer = IndirectActorProducer` and `TypedCreatorFunctionConsumer` is an IndirectActorProducer

will this change anything in when the  `behavior()` is called? timing? thread?

Reply:
It used to be the `Props#apply(creator: => T)` factory, which already was leading to the `TypedCreatorFunctionConsumer` path (but now avoiding an args-to-varargs-to-list dance by calling the constructor right away).

Reply:
right, `: =>`